### ISSUE TITLE
How the Input tensor definition for tensorflow GraphDef model run in TRTIS?

### ISSUE THREAD

**Issue by ivankxt** [NVIDIA/tensorrt-inference-server#33]
The tensorflow GraphDef model have two inputs,  one is input_a and  the other is input_b. In the graph.pbtxt file, the input_a is defined as:
value {
      list {
        shape {
          dim {
            size: -1
          }
        }
      }
    }
, and the input_b is defined as:
value {
      list {
        shape {
          dim {
            size: -1
          }
          dim {
            size: -1
          }
        }
      }
    }
, now, I want to konw how the dims definition of config.pbtxt file for input_a and input_b respectively ?

**Comment by deadeyegoodwin** [comment]
Shapes:
input_a = [ -1 ]
input_b = [ -1, -1 ]

The -1 means that any value is allowed for that dimension. These "variable-sized" dimensions are supported by TF but not yet by TRTIS (but we are actively working on it and plan to have it available in the next release, see also #8.

You can still support a specific size by specifying that size in the model configuration file, config.pbtxt. For example, if your input_a is always size [10] and your input_b is always size [100,100], just use those dimensions in the model configuration.

The leading -1 may also indicate the batch dimension. Batching is supported by TRTIS but in the model configuration file you don't list the batch dimension. There is discussion of this in the model configuration documentation: https://docs.nvidia.com/deeplearning/sdk/tensorrt-inference-server-guide/docs/model_configuration.html#

**Comment by ivankxt** [comment]
Hello, @deadeyegoodwin 

My tensorflow GraphDef model inputs and outputs defined in graph.pbtxt file as:
inputs: input_a0, input_a1, ..., input_a44 dims are [ -1 ] and input_b0,..., input_b3 dims are  [ -1, -1 ],
outputs: test_mlp/logits/dense/BiasAdd dims is [ -1, 11]

Loading the tensorflow GraphDef  model by session run, we got its output result as:
 [-0.187565967 -0.0866315216 -0.162302822 -0.00336572342 -0.207192868 -0.046942886 -0.0337462276 -0.437082946 -0.641982377 -0.0126479603 -0.0623684227]

However, when loading the tensorflow GraphDef  model by TRTIS, we got the error as:
File "/usr/local/lib/python2.7/dist-packages/tensorrtserver/api/__init__.py", line 190, in _raise_if_error
    raise ex
tensorrtserver.api.InferenceServerException: [inference:0 8] In[0] is not a matrix
         [[Node: test_mlp/hiddenlayer_0/dense/MatMul = MatMul[T=DT_FLOAT, _output_shapes=[[?,48]], transpose_a=false, transpose_b=false, _device="/job:localhost/replica:0/task:0/device:GPU:0"](test_mlpinput/Elu, test_mlp/hiddenlayer_0/dense/kernel/read)]]
         [[Node: test_mlp/logits/dense/BiasAdd/_99 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device_incarnation=1, tensor_name="edge_466_test_mlp/logits/dense/BiasAdd", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]

Here our config.pbtxt file as:
name: "test_graphdef"
platform: "tensorflow_graphdef"
max_batch_size: 100
input [
  {
    name: "input_a0"
    data_type: TYPE_INT64
    dims: [ 1 ]
  },
  {
    name: "input_a1"
    data_type: TYPE_INT64
    dims: [ 1 ]
  },
......
  {
    name: "input_a44"
    data_type: TYPE_INT64
    dims: [ 1 ]
  },
  {
    name: "input_b0"
    data_type: TYPE_INT64
    dims: [ 1, 1 ]
  },
......
  {
    name: "input_b3"
    data_type: TYPE_INT64
    dims: [ 1, 1 ]
  }
]
output [
  {
    name: "test_mlp/logits/dense/BiasAdd"
    data_type: TYPE_FP32
    dims: [ 1, 11 ]
  }
]

And the python code as:
batch_size = 1
input1_data = np.arange(start=1, stop=2, dtype=np.int64)
input2_data = np.matlib.ones((1, 1),dtype=np.int64)

result = ctx.run({ 
                       'input_a0' : (input1_data,)
                       ,'input_a1' : (input1_data,)
                       ......
                       ,'input_a44' : (input1_data,)

                       ,'input_b0' : (input2_data,) 
                       ......
                       ,'input_b3' : (input2_data,) 
                       },
                     { 'test_mlp/logits/dense/BiasAdd' : InferContext.ResultFormat.RAW},
                     batch_size)

Could you give me some explanation for that? Thank You.

**Comment by ivankxt** [comment]
> Shapes:
> input_a = [ -1 ]
> input_b = [ -1, -1 ]
> 
> The -1 means that any value is allowed for that dimension. These "variable-sized" dimensions are supported by TF but not yet by TRTIS (but we are actively working on it and plan to have it available in the next release, see also #8.
> 
> You can still support a specific size by specifying that size in the model configuration file, config.pbtxt. For example, if your input_a is always size [10] and your input_b is always size [100,100], just use those dimensions in the model configuration.
> 
> The leading -1 may also indicate the batch dimension. Batching is supported by TRTIS but in the model configuration file you don't list the batch dimension. There is discussion of this in the model configuration documentation: https://docs.nvidia.com/deeplearning/sdk/tensorrt-inference-server-guide/docs/model_configuration.html#

Hello, deadeyegoodwin

My tensorflow GraphDef model inputs and outputs defined in graph.pbtxt file as:
inputs: input_a0, input_a1, ..., input_a44 dims are [ -1 ] and input_b0,..., input_b3 dims are [ -1, -1 ],
outputs: test_mlp/logits/dense/BiasAdd dims is [ -1, 11]

Loading the tensorflow GraphDef model by session run, we got its output result as:
[-0.187565967 -0.0866315216 -0.162302822 -0.00336572342 -0.207192868 -0.046942886 -0.0337462276 -0.437082946 -0.641982377 -0.0126479603 -0.0623684227]

However, when loading the tensorflow GraphDef model by TRTIS, we got the error as:
File "/usr/local/lib/python2.7/dist-packages/tensorrtserver/api/init.py", line 190, in _raise_if_error
raise ex
tensorrtserver.api.InferenceServerException: [inference:0 8] In[0] is not a matrix
[[Node: test_mlp/hiddenlayer_0/dense/MatMul = MatMul[T=DT_FLOAT, _output_shapes=[[?,48]], transpose_a=false, transpose_b=false, _device="/job:localhost/replica:0/task:0/device:GPU:0"](test_mlpinput/Elu, test_mlp/hiddenlayer_0/dense/kernel/read)]]
[[Node: test_mlp/logits/dense/BiasAdd/_99 = _Recvclient_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device_incarnation=1, tensor_name="edge_466_test_mlp/logits/dense/BiasAdd", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"]]

Here our config.pbtxt file as:
name: "test_graphdef"
platform: "tensorflow_graphdef"
max_batch_size: 100
input [
{
name: "input_a0"
data_type: TYPE_INT64
dims: [ 1 ]
},
{
name: "input_a1"
data_type: TYPE_INT64
dims: [ 1 ]
},
......
{
name: "input_a44"
data_type: TYPE_INT64
dims: [ 1 ]
},
{
name: "input_b0"
data_type: TYPE_INT64
dims: [ 1, 1 ]
},
......
{
name: "input_b3"
data_type: TYPE_INT64
dims: [ 1, 1 ]
}
]
output [
{
name: "test_mlp/logits/dense/BiasAdd"
data_type: TYPE_FP32
dims: [ 1, 11 ]
}
]

And the python code as:
batch_size = 1
input1_data = np.arange(start=1, stop=2, dtype=np.int64)
input2_data = np.matlib.ones((1, 1),dtype=np.int64)

result = ctx.run({
'input_a0' : (input1_data,)
,'input_a1' : (input1_data,)
......
,'input_a44' : (input1_data,)
,'input_b0' : (input2_data,)
......
,'input_b3' : (input2_data,)
},
{ 'test_mlp/logits/dense/BiasAdd' : InferContext.ResultFormat.RAW},
batch_size)

Could you give me some explanation for that? Thank you.

**Comment by deadeyegoodwin** [comment]
When your configuration says this:

max_batch_size: 100
input [
{
name: "input_a0"
data_type: TYPE_INT64
dims: [ 1 ]
},

That means that TRTIS will expect the model to be able to accept an input_a0 tensor with shape [ 100, 1 ] (or shapes [ 1, 1 ], [ 2, 1 ], ... [ 99, 1 ], [ 100, 1] since the first dimension is the batch dimension and supports up to max_batch_size). You do not specify the batch dimension in the model configuration (this is discussed in the documentation pointer in the prior response).

Your model accepts input_a0 as [ -1 ], so there is a mismatch. As mentioned earlier what does that -1 dimension represent. If it is the batch dimension then input_a0 is actually dimensionless in the model configuration (which is currently not supported). If your model doesn't support batching and so input_a0 is just an arbitrary length 1-dimensional vector.... then you need to set max_batch_size: 0 to indicate that your model doesn't support batching.
